File: JS/LinkedList/LinkedList.js
/*
  Main approaches to solve LinkedList question are using recursion or 2 pointers approach
*/
class Node {
  data;
  next;

  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  head;
  tail;
  length = 0;

  constructor(data) {
    let newNode = new Node(data);
    this.head = newNode;
    this.tail = newNode;
    this.length++;
  }

  append = (data) => {
    let newNode = new Node(data);
    this.tail.next = newNode;
    this.tail = newNode;
    this.length++;
  };

  prepend = (data) => {
    let newNode = new Node(data);
    newNode.next = this.head;
    this.head = newNode;
    this.length++;
  };

  deleteNode = (data) => {
    let prev = null,
      current = this.head;
    let isNodePresent = true;
    while (current.data != data && current != null) {
      prev = current;
      current = current.next;
      if (current == null) {
        isNodePresent = false;
      }
    }
    if (!isNodePresent) {
      console.log("This node is not present in LinkedList");
    } else {
      prev.next = current.next;
      this.length--;
    }
  };

  isListEmpty = () => {
    return this.length ? false : true;
  };

  printList = () => {
    console.log("List print starts from here");
    let current = this.head;
    let printData = "";
    while (current != null) {
      printData += current.data + " ";
      current = current.next;
    }
    console.log(printData);
  };
}

module.exports = { LinkedList, Node };

File: Java/LinkedList/LinkedList.java
package Java.LinkedList;

public class LinkedList<T> {
    Node<T> head;
    Node<T> tail;
    int length;

    LinkedList(T data) {
        Node<T> newNode = new Node<>(data);
        this.head = newNode;
        this.tail = newNode;
        this.length = 0;
    }

    public void append(T data) {
        Node<T> newNode = new Node<>(data);
        this.tail.next = newNode;
        this.tail = newNode;
        this.length++;
    }

    public void prepend(T data) {
        Node<T> newNode = new Node<>(data);
        newNode.next = this.head;
        this.head = newNode;
        this.length++;
    }

    public void deleteNode(T data) {
        Node<T> prev = null, current = this.head;
        boolean isNodePresent = true;
        while (current.data != data && current != null) {
            prev = current;
            current = current.next;
            if (current == null) {
                isNodePresent = false;
            }
        }
        if (!isNodePresent) {
            System.out.println("Node with this data is not present");
        } else {
            prev.next = current.next;
            this.length--;
        }
    }

    public boolean isListEmpty() {
        return this.length > 0 ? false : true;
    }

    public void printList() {
        Node<T> current = this.head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }

}
